helix-term/src/compositor.rs

```diff
@@ -243,6 +243,10 @@ impl Compositor {
             .find(|component| component.id() == Some(id))
             .and_then(|component| component.as_any_mut().downcast_mut())
     }
+
+    pub fn has_overlay(&self) -> bool {
+        self.layers.len() > 1
+    }
```

Comment:
This feels like code smell, but I don't know the right way to think about this. If we don't check to see if there's an overlay, then auto-hover will paint on top of any active pickers. Is there a better way to check for multiple layers?

Reply:
There's a `has_component` function. You should probably check for pickers instead of overlay though.

Reply:
IIRC you have to check for an `Overlay<Picker>` type instead of a `Picker` type with `has_component()` since pickers area always wrapped in an overlay.

Reply:
Thank you both! 🙂

Reply:
Ok, did an experiment with this. It may not be feasible to use `has_component` to check for an occluding layer without enumerating every possible `Picker`. I did a quick check on `type_name()`s to verify if you could look for `Overlay<Picker>` and it looks like it would have to be the fully-qualified, fully-specialized names, e.g.:

```
// for a file picker
helix_term::ui::overlay::Overlay<helix_term::ui::picker::FilePicker<std::path::PathBuf>>

// for a symbol picker
helix_term::ui::overlay::Overlay<helix_term::ui::picker::FilePicker<lsp_types::document_symbols::SymbolInformation>>
```

Reply:
You could implement `id()` for `Overlay` and `Picker` and write a `has_id()` function. The `Overlay` id could be `format!("Overlay<{}>", T::id())`, and the `Picker` id could just be `"Picker"`. Alternatively, you could have `Overlay` just return `"Overlay"`.

Reply:
`Component::id()` is supposed to uniquely identify the component if supplied, so this might not be a good idea. Maybe there can be a variant of `has_component()` that uses a predicate:

```rust
    pub fn has_component_by(&self, predicate: impl Fn(&str) -> bool) -> bool {
        self.layers
            .iter()
            .map(|component| component.type_name())
            .any(predicate)
    }
```

and you would use it like `compositor.has_component_by(|c| c.contains("Picker"))`. Well TBH this is a bit hacky too :)

Reply:
I do like that slightly more--it's at least more generally useful than `has_overlay`. I'll give it a shot and push it up here.